azure-kusto-ingest/azure/kusto/ingest/_retry.py

```diff
@@ -11,8 +16,28 @@ def __init__(self, max_attempts, sleep_base_secs: float = 1.0, max_jitter_secs:
         self.current_attempt = 0
 
     def do_backoff(self):
-        sleep((self.sleep_base_secs * (2**self.current_attempt)) + (random.uniform(0, self.max_jitter_secs)))
+        """
+        Update retry counter and sleep before the next retry.
+        """
         self.current_attempt += 1
+        if self:
+            sleep((self.sleep_base_secs * (2**self.current_attempt)) + (random.uniform(0, self.max_jitter_secs)))
 
     def __bool__(self):
         return self.current_attempt < self.max_attempts
+
+
+def func_retry_wrapper(func, max_retry, sleep_base_secs: float, *exceptions):
```

Comment:
Ideally I'd like it to be a decorator, but because it receives parameters that are dependent on class members that's hard.

Do you have an idea as to how or do you prefer it like this?

Reply:
I was also thinking about decorator but than the "retry scope" is the entire decorated function (even though you may only want to retry a specific part or line). I feel like this is more practical. I agree that decorator is more appealing.

Reply:
Correct me if I'm wrong, but this should be less of a problem since we only retry on the specific exception.
So even if we cover "too much" in the retry we won't accidentally retry something we don't

Reply:
You are correct but there might be more complex scenarios in the future and this approach is more flexible.
I don't mind adding a decorator and let the developer choose as he sees fit. Are you OK with having both options (decorator | direct call to 'func_retry_wrapper')?

Reply:
The python zen is "There should be one-- and preferably only one --obvious way to do it.", I don't think we should put code in if we're not going to use it.

A decorator is just as flexible, you can call it like this if you want
```python
my_decorator(partial(self._kusto_client.execute, "NetDefaultDB", ".get ingestion resources"))()
```